### DISCUSSION TITLE
Training suddenly stopped working correctly: dreamy, grayscale, incoherent output images

### DISCUSSION THREAD

**Discussion by minienglish1** [d8ahazard/sd_dreambooth_extension#969] [General]
Initializing Dreambooth
If submitting an issue on github, please provide the below text for debugging purposes:

Python revision: 3.10.6 (tags/v3.10.6:9c7b4bd, Aug  1 2022, 21:53:49) [MSC v.1932 64 bit (AMD64)]
Dreambooth revision: 43ae9d55531004f1dedaea7ac2443e9b16739913
SD-WebUI revision: 0cc0ee1bcb4c24a8c9715f66cede06601bfc00c8

Checking Dreambooth requirements...
[+] bitsandbytes version 0.35.0 installed.
[+] diffusers version 0.10.2 installed.
[+] transformers version 4.25.1 installed.
[+] xformers version 0.0.16rc425 installed.
[+] torch version 1.13.1+cu117 installed.
[+] torchvision version 0.14.1+cu117 installed.
-----------------------------------------------

With that out of the way, I few days ago I was continuing my a/b testing, when my output sample images became very dreamy, formless, and lots of grayscale images.  I blamed it on the a/b test and some new images I added to my data set.  Turns out, nothing properly trained after that.  I probably updated the WebUI and/or dreambooth around that time, too.  But I don't see anyone else having issues, so I'm doubt it was the update.

Parameters used while trying to troubleshoot the problem:
GPU: RTX 3090
Training parameters: 2.1 768 non-ema base ckpt, 768 resolution, unfreeze model, batch size 1, FP16, 8bit Adam, xformers, use ema, cache latents, train Unet, Step Ratio of Text Encoder Training 1, Freeze CLIP, Pad Tokens.  Data set:10 photos of a girl in a park with prompts, 1e-5 learning rate, 100 epochs, instance prompt: [filewords], fine-tuning - not dreambooth training.

These are all parameters and a dataset I've used many many times before without problem.

I've since then done the following to try to fix it:
1) Completely deleted everything (including pip cache and models/yaml) and re-downloaded and re-installed
2) Using flash attention instead of xformers (doesn't go all dreamy, but it also doesn't capture the subject correctly)
3) Trained a short 25 epochs in every dream 2 trainer to see if it was a system problem (training came back good, so most likely not a system-wide issue. doing a longer training now. )
4) Changing to safetensors version of base model.  (no change)
5) Upgrading xformers to latest dev build (improvement, but background gets blended into the person, pictures look painted - not like a photo )

By accident, after testing upgrading xformers, I generated an picture without a prompt.  The image came back near perfect, close to what it would have been before this problem.  It's like the images are being trained, but the caption are only partially trained.  So I started messing with caption settings: max token lengths of 75/150, and shuffle tokens.  But no luck.

So, it seems like the CLIP part is acting weird.  I'm not sure what to make of it, or what steps to take next.  I plan to try installing xformers 0.0.14, since that's what every dream 2 trainer uses.  Also could try a different base mode, but I only train 768 resolution and up. So if that works, it doesn't fix my problem.  Past that, I'm at a loss what to do.

Anyone have any suggestions? 

Copy of the last troubleshooting json config:

    "adamw_weight_decay": 0.01,
    "attention": "xformers",
    "cache_latents": true,
    "center_crop": false,
    "freeze_clip_normalization": true,
    "clip_skip": 1,
    "concepts_list": [
        {
            "class_data_dir": "C:\\Users\\3090\\stable-diffusion-webui\\models\\dreambooth\\TEST-Newxformers\\classifiers_0",
            "class_guidance_scale": 7.5,
            "class_infer_steps": 40,
            "class_negative_prompt": "",
            "class_prompt": "",
            "class_token": "",
            "instance_data_dir": "c:\\test",
            "instance_prompt": "",
            "instance_token": "[filewords]",
            "is_valid": true,
            "n_save_sample": 0,
            "num_class_images": 0,
            "num_class_images_per": 0,
            "sample_seed": -1,
            "save_guidance_scale": 7.5,
            "save_infer_steps": 40,
            "save_sample_negative_prompt": "",
            "save_sample_prompt": "",
            "save_sample_template": ""
        }
    ],
    "concepts_path": "",
    "custom_model_name": "",
    "epoch": 100,
    "epoch_pause_frequency": 0.0,
    "epoch_pause_time": 0.0,
    "gradient_accumulation_steps": 1,
    "gradient_checkpointing": false,
    "gradient_set_to_none": true,
    "graph_smoothing": 50.0,
    "half_model": false,
    "train_unfrozen": true,
    "has_ema": false,
    "hflip": false,
    "initial_revision": 0,
    "learning_rate": 1e-06,
    "learning_rate_min": 1e-06,
    "lifetime_revision": 0,
    "lora_learning_rate": 0.0002,
    "lora_model_name": [],
    "lora_rank": 4,
    "lora_txt_learning_rate": 0.0002,
    "lora_txt_weight": 1,
    "lora_weight": 1,
    "lr_cycles": 1,
    "lr_factor": 0.5,
    "lr_power": 1,
    "lr_scale_pos": 0.5,
    "lr_scheduler": "constant_with_warmup",
    "lr_warmup_steps": 0,
    "max_token_length": 150,
    "mixed_precision": "fp16",
    "model_name": "TEST-Newxformers-1e-6-150-shuffle",
    "model_dir": "C:\\Users\\3090\\stable-diffusion-webui\\models\\dreambooth\\TEST-Newxformers-1e-6-150-shuffle",
    "model_path": "C:\\Users\\3090\\stable-diffusion-webui\\models\\dreambooth\\TEST-Newxformers-1e-6-150-shuffle",
    "num_train_epochs": 100,
    "pad_tokens": true,
    "pretrained_model_name_or_path": "C:\\Users\\3090\\stable-diffusion-webui\\models\\dreambooth\\TEST-Newxformers-1e-6-150-shuffle\\working",
    "pretrained_vae_name_or_path": null,
    "prior_loss_scale": false,
    "prior_loss_target": 100.0,
    "prior_loss_weight": 0.75,
    "prior_loss_weight_min": 0.1,
    "resolution": 768,
    "revision": 1000,
    "sample_batch_size": 1,
    "sanity_prompt": "",
    "sanity_seed": 420420.0,
    "save_ckpt_after": true,
    "save_ckpt_cancel": false,
    "save_ckpt_during": true,
    "save_embedding_every": 25,
    "save_lora_after": true,
    "save_lora_cancel": false,
    "save_lora_during": false,
    "save_preview_every": 0,
    "save_safetensors": true,
    "save_state_after": false,
    "save_state_cancel": false,
    "save_state_during": false,
    "scheduler": "ddim",
    "shuffle_tags": false,
    "snapshot": "",
    "src": "C:\\Users\\3090\\stable-diffusion-webui\\models\\Stable-diffusion\\v2-1_768-nonema-pruned.safetensors",
    "stop_text_encoder": 1,
    "train_batch_size": 1,
    "train_imagic": false,
    "train_unet": true,
    "use_8bit_adam": true,
    "use_concepts": false,
    "use_ema": true,
    "use_lora": null,
    "use_subdir": true,
    "v2": true
}

**Comment by Arilziem** [comment]
So a couple of thoughts:
* I think your main problem is that you are using "instance token" instead of "instance prompt". That would lead to your overall model slowly overfitting to the imagery, but without really learning the subject. I always get those two mixed up as well, but iirc the tokens are just convenience functions that add/replace stuff in your prompts. So use `[filewords]` as your `instance prompt` instead and you should get vastly better results.
* definitely use either xformers 0.14 or 0.17 (and not 0.16), with 0.17 being the faster alternative afaik
* You may want to reduce the text encoder ratio, at least for me it tends to overfit quickly

**Reply by minienglish1** [reply]
Thanks so much.  I think you have found the problem  It wasn't intentional to use "instance token."  I've trained so many models that filling out the GUI is done without thinking, much less even reading the fields.  

I usually just load the settings from the last model and make adjustments.  Occasionally I clear out all my testing models and start from scratch.  If when I did that, I put "instance prompt" in the "instance token field," and just modified the settings from there on the future models, that would explain everything.  That's what I get for working on auto-pilot.   Good to know it's just a stupid simple mistake.

I'm still training a model on every dream 2 trainer, when that finishes in the morning, I test to see if changing to "instance prompt" works.  99% sure it will.

As far as xformers 0.16, it's functional, at least on my gpu.  I've used previously for a good while.  That's why I used it to troubleshoot, because it is know to work correctly in the past.  But I definitely will update to 0.17 (and probably update cudnn too) once I get this mess sorted out.

Again, thanks!

**Reply by Arilziem** [reply]
Yeah I did the same thing a few weeks ago, kinda sounded familiar lol

**Reply by minienglish1** [reply]
Also, regarding the text encoder ratio:  Isn't the text encoder ratio the same as conditional dropout?  I plan to A/B test it once I settle on a learning rate and FP16vsFP32 for my data set. 

What text encoder ratio do you use for fine-tuning?

**Reply by Arilziem** [reply]
I don't think they are the same, no. The text encoder ratio determines how many steps compared to the UNET are used to train the CLIP model. Conditional dropout just skips the captions during UNET training sometimes, no? I don't think that's even implemented yet.

I haven't experimented much with the text encoder yet since I only have a 3080 and separate training of CLIP wasn't a thing until recently (can't do both with my 10GB VRAM). I got decent results with what amounts to around 0.3 iirc. 0.5 is probably fine as well. Nice thing is that you can easily swap out the CLIP, make backups etc. without too many negative effects from what I can tell. So if you overtrain you should be able to just swap it out with a backup and then retrain only CLIP.

**Reply by minienglish1** [reply]
Glad I asked.  Some trainers have conditional dropout enabled.  I'll play around with it next week.

How do you swap out the CLIP and and make backups?  Train each separately and combine them or what?  I've always just trained them together.  Didn't even think about training them separately.

Also, just ran 50 epochs on a 391pic data set, everything looks great.  Thanks for the help finding the error.

**Reply by Arilziem** [reply]
In `models/dreambooth/<your model>/working/CLIP` (I think) you can find a model file that you can copy or swap out. It will be used in subsequent trainings and when compiling checkpoints.

**Reply by minienglish1** [reply]
Makes sense.  Thanks.

**Comment by saunderez** [comment]
Instance and class tokens work fine. Used in conjunction with filewords they help the model better isolate your subject from the traits it shares with other subjects in the same class. Its when you use a static instance prompt that tends to overfit because you show the model a bunch of really similar images and tell it they're all the same thing and  then you reward it coz the images are low loss. And the reason the loss is so low is because the images get closer and closer to the dataset. Until they all look like the dataset.

**Reply by Arilziem** [reply]
Yes, they work fine. But they only work in conjunction with the prompts (except for imagic I think?).

Tokens are used to dynamically build prompts, which is mainly used for generating the preservation images. For example, if you have the prompt `"a dude in a cave"`, an `instance_token` `"dude"` and a class_token `"man"`, it will replace `"dude"` (instance) with `"man"` (class) when generating the class preservation images (result: "a man in a cave").

Also, it will do the opposite when not generating class images (as far as I can tell also for the actual training), replacing every "man" with a "dude" if there is a man found in the prompt. For cases where neither instance_token nor class_token are found in the prompt, the final (non-class) prompt will become `[instance_token] [class_token], [prompt]`.

Putting `[filewords]` into the `instance_token` doubly doesn't work, since the content of `instance_token` isn't interpreted (as opposed to the prompts).

In @minienglish1's case, the entire training was effectively unconditional training, because there was no `class_token`, in which case `instance_token` is just ignored and so the `instance_prompt` was used verbatim (but it was empty).

For cases when you have no prompt, but do have `instance_token` and `class_token`, the resulting prompt will become `[instance_token] [class_token]`. So yes, that will work (somewhat) fine even without a base prompt as long as you fill in both tokens. Not advisable for finetuning though.

**Comment by 78Alpha** [comment]
I see you are using `[filewords]` and I have the same issue with it. If you go to the folder containing your samples, you will find an empty text file that would have contained the prompt in `[filewords]` but it will be 0 Bytes and blank. If you edit the image in something like notepad++, you will find the prompt field completely missing from the image. Seems to be an issue with how it handles filewords now. In older versions it worked, but sometime in the past 1 or 2 months it has had the issue.

If you manually input a sample prompt, it should work no problem.

I will add an edit. You may also want to view extensions on your files, make sure they didn't get any hidden things added like a `filename.jpg.txt` that looks like `filename.txt`

**Reply by minienglish1** [reply]
Thanks.  I figured it out.  Also the Dataset Tag Editor will blank the caption.txt file if you manually edit one file in it.  Notepad++ seems to work fine for me, and is my go to for manually editing 1 caption.
